### ISSUE TITLE
[10.0.0-beta.6] vscode react native debugger failed to run and reporting require cycle

### ISSUE THREAD

**Issue by deckyfx** [realm/realm-js#3132]
Realm version: 10.0.0-beta.6

But I think this problem should also occur in version: 10.0.0-beta.9

this is related to [#3085](https://github.com/realm/realm-js/issues/3085)

when running debugger tool in vs code the app will fail to run and the debug console will report this

The debugger in vscode also reporting this

```
Require cycle: node_modules/realm/lib/browser/util.js -> node_modules/realm/lib/browser/rpc.js -> node_modules/realm/lib/browser/util.js
Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
YellowBox.js:71
ReferenceError: createSession is not defined
ExceptionsManager.js:44
Running "SHARK" with {"rootTag":11}
infoLog.js:16
Invariant Violation: "SHARK" has not been registered. This can happen if:
* Metro (the local dev server) is run from the wrong folder. Check if Metro is running, stop it and restart it in the current project.
* A module failed to load due to an error and `AppRegistry.registerComponent` wasn't called.
```

and this is how I bypass both issues, there are several files you need to modify, the base path is

```<your_project>/node_modules/realm/```

in ```lib/browser/index.js```
add this import
```import {createSession} from './session';```
around line 213, change the block inside try into this
```
let refreshAccessTokenCallback = staticUserMethods && staticUserMethods._refreshAccessToken ? staticUserMethods._refreshAccessToken.bind(User) : () => {};
// The session ID refers to the Realm constructor object in the RPC server.
Realm[keys.id] = createSession(refreshAccessTokenCallback, debugHosts[i] + ':' + debugPort);
```

create new file ```lib/browser/cache.js```
```
export let propertyCache = {};

export function invalidateCache(realmId) {
    if (realmId) {
        propertyCache[realmId] = {};
    }
    else {
        propertyCache = {};
    }
}

export function getRealmCache(realmId) {
    let realmCache = propertyCache[realmId];
    if (!realmCache) {
        realmCache = propertyCache[realmId] = {};
    }
    return realmCache;
}
```

in ```lib/browser/util.js```

add this import
```
import {invalidateCache, getRealmCache} from './cache'
```
then remove declaration of variable ```propertyCache``` function ```invalidateCache``` and ```getRealmCache``` around  line 62

in ```lib/browser/rpc.js```
add this import
```
import {invalidateCache} from './cache'
```

At line 64, change
```
util.invalidateCache(realm[keys.realm]);
```
into
```
invalidateCache(realm[keys.realm]);
```

And finally in ```lib/browser/session.js```
around line 56, wrap the line with if
```
if (info && info.data && info.data.user) {
        sessionProxy[keys.id] = info.id;
        sessionProxy[keys.type] = objectTypes.SESSION;
        sessionProxy.user = deserialize(realmId, info.data.user);
        sessionProxy.config = deserialize(realmId, info.data.config)
    }
```

Also the debugger may also require your PC and Phone to be connected in the same WIFI

In short, 
I put some safety check against null or undefined variable so it does not raise errors when executed
Also should separate cache function and import only the necessary function to avoid cyclic require that I believe should improve performance

**Comment by deckyfx** [comment]
Fixed in #3135

**Comment by ferbeb** [comment]
I'm afraid a typo made it into in the import in index.js, breaking the beta 10 released today:
`import { createSession } './session';`
The _from_ is missing.
